src/include/serial_inline.h

```diff
@@ -63,7 +72,14 @@ __insert_serial_func(WT_SESSION_IMPL *session, WT_INSERT_HEAD *ins_head, WT_INSE
      * implementations read the old value multiple times.
      */
     for (i = 0; i < skipdepth; i++) {
-        WT_INSERT *old_ins = *ins_stack[i];
+        /*
+         * We assume the value of old_ins would not change within this loop. However, the compiler
+         * may change the code in a way that it will read the old_ins from memory again.
+         *
+         * Place a compiler barrier here to avoid this issue.
+         */
+        old_ins = *ins_stack[i];
+        WT_BARRIER();
```

Comment:
This will introduce a hardware barrier to the code. While my understanding is that we only need a compiler barrier here as we have a cas later which already functions as a hardware barrier.

Reply:
I think a barrier is required here. If the CPU/hardware decided to optimize this, it would have the same consequence of producing incorrect results.

I'd argue that we should look very carefully at every place we have a plain compiler barrier - it's unlikely to be strong enough to do what we require. That falls into future work IMO

Reply:
I still don't understand how CPU can optimize this. My understanding is that CPU can only reorder reads but not introducing new reads. Since we have a cas later, all the reads should be already ordered in CPU.